src/freenet/clients/fcp/FCPPluginClient.java

```diff
@@ -0,0 +1,1190 @@
+     *             FCPPluginMessage, long)} which will return the reply right away.</p>
+     * @see #sendSynchronous(SendDirection, FCPPluginMessage, long)
+     *          You may instead use the blocking sendSynchronous() if your thread needs to know
+     *          whether messages arrived, to ensure a certain order of arrival, or to know
+     *          the reply to a message.
+     */
+    public void send(final SendDirection direction, FCPPluginMessage message) throws IOException {
+        // We first have to compute the message.permissions field ourselves - we shall ignore what
+        // caller said for security.
+        ClientPermissions currentPermissions = (direction == SendDirection.ToClient) ?
+             null // Server-to-client messages do not have permissions.
+             : computePermissions();
+
+        // We set the permissions by creating a fresh FCPPluginMessage object so the caller cannot
+        // overwrite what we compute.
+        message = FCPPluginMessage.constructRawMessage(currentPermissions, message.identifier,
+            message.params, message.data, message.success, message.errorCode, message.errorMessage);
+
+        // Now that the message is completely initialized, we can dump it to the logfile.
+        if(logDEBUG) {
+            Logger.debug(this, "send(): direction = " + direction + "; " + "message = " + message);
+        }
+        
+        // True if the target server or client message handler is running in this VM.
+        // This means that we can call its message handling function in a thread instead of
+        // sending a message over the network.
+        // Notice that we do not check for server != null because that is not allowed by this class.
+        final boolean messageHandlerExistsLocally =
+            (direction == SendDirection.ToServer) ||
+            (direction == SendDirection.ToClient && client != null);
+        
+        if(!messageHandlerExistsLocally) {
+            dispatchMessageByNetwork(direction, message);
+            return;
+        }
+        
+        assert(direction == SendDirection.ToServer ? server != null : client != null)
+            : "We already decided that the message handler exists locally. "
+            + "We should have only decided so if the handler is not null.";
+        
+        // Since the message handler is determined to be local at this point, we now must check
+        // whether it is a blocking sendSynchronous() thread instead of a regular
+        // FredPluginFCPMessageHandler.
+        // sendSynchronous() does the following: It sends a message and then blocks its thread
+        // waiting for a message replying to it to arrive so it can return it to the caller.
+        // If the message we are processing here is a reply, it might be the one which a
+        // sendSynchronous() is waiting for.
+        if(maybeDispatchMessageLocallyToSendSynchronousThread(direction, message))
```

Comment:
Commit df8e981e8e359d2fec7ad3af0e1b18c1c4e767e8 improves the preceding comment to better explain the "when?".
It also renames the function to be more self-explanatory.

The new name is even longer though :D This is because I have two policies for function name length, depending on the type of function:
1. Functions for "daily use", i.e. which will be used frequently in new code, shall have short names. The perfect example is the send() function which we are discussing here. send()ing network messages is something which developers will do frequently in new code, and thus the name is very short.
2. One-shot functions / internal code. This is the case with the maybe...() function whose name you wanted to shorten. It only has *one* caller; it is private; and it only exists to split up send(), not to provide any API for future use. So as its unlikely that anyone will ever have to type its name again, the name being self-explanatory is more important than saving keystrokes.

I hope you can agree with the reasoning with regards to name length.